pkg/requestor/requestor.go

```diff
@@ -83,51 +85,73 @@ func (r *Requestor) consumeMessages() error {
-				bm := messages.NewBuildMessage(-1)
-				err1 = json.Unmarshal(d.Body, bm)
-				klog.V(4).Infof("received build message %#v", bm)
-				if err1 != nil {
-					done <- fmt.Errorf("failed to unmarshal as build message %w", err1)
-					return
-				}
-				r.jenkinsBuild = bm.Build
-				fmt.Printf("Following jenkins build %d\n", r.jenkinsBuild)
-			} else if r.jenkinsBuild == m.Build {
-				if m.ISLog() {
-					klog.V(2).Infof("received log message")
-					lm := messages.NewLogsMessage(-1, "")
-					err1 = json.Unmarshal(d.Body, lm)
-					if err1 != nil {
-						done <- fmt.Errorf("failed to unmarshal as logs message %w", err1)
-						return
-					}
-					klog.V(4).Infof("log message %#v", lm)
-					fmt.Println(lm.Logs)
-				} else if m.IsStatus() {
-					klog.V(2).Infof("received status message")
-					sm := messages.NewStatusMessage(-1, false)
-					err1 = json.Unmarshal(d.Body, sm)
-					if err1 != nil {
-						done <- fmt.Errorf("failed to unmarshal as status message %w", err1)
+			//process message only if the jenkins projects match
+			if m.JenkinsProject == r.jenkinsProject {
+				if m.Build > r.jenkinsBuild {
+					//process build or cancel message, only if the message build > current jenkins build
+					if m.IsBuild() {
+						klog.V(2).Infof("received build message")
+						bm := messages.NewBuildMessage(-1,"")
+						err1 = json.Unmarshal(d.Body, bm)
+						klog.V(4).Infof("received build message %#v", bm)
+						if err1 != nil {
+							done <- fmt.Errorf("failed to unmarshal as build message %w", err1)
+							return
+						}
+						r.jenkinsBuild = bm.Build
+						fmt.Printf("Following jenkins build %d\n", r.jenkinsBuild)
+					} else if m.IsCancel() {
+						klog.V(2).Infof("received cancel message from newer build")
+						cm := messages.NewCancelMessage(-1, "")
+						err1 := json.Unmarshal(d.Body, cm)
+						if err1 != nil {
+							done <- fmt.Errorf("failed to unmarshal cancel message %w", err1)
+						}
+						fmt.Printf("\n!!! Detected newer build %d and following it. Please ignore logs above this point !!!\n", cm.Build)
+						r.jenkinsBuild = -1
+						success = true
 					}
-					klog.V(4).Infof("status message %#v", sm)
-					if success {
-						klog.V(2).Infof("Updating success status")
-						success = sm.Success
-					} else {
-						klog.V(2).Infof("Skipping update to status as its already false")
-					}
-				} else if m.IsFinal() {
-					klog.V(2).Infof("received final message")
-					if success {
-						done <- nil
+				} else if r.jenkinsBuild == m.Build {
+					//process other types of messages, only if message build matches current jenkins build
+					if m.IsLog() {
+						klog.V(2).Infof("received log message")
+						lm := messages.NewLogsMessage(-1, "", "")
+						err1 = json.Unmarshal(d.Body, lm)
+						if err1 != nil {
+							done <- fmt.Errorf("failed to unmarshal as logs message %w", err1)
+							return
+						}
+						klog.V(4).Infof("log message %#v", lm)
+						fmt.Println(lm.Logs)
```

Comment:
Too many nested loop. Can't we customise it further.

Reply:
I dont think this is a nested loop at all. There is one loop that processes amqp deliveries. You could argue of course that there a depth in conditionals, which could maybe be optimized.

Reply:
> You could argue of course that there a depth in conditionals, which could maybe be optimized.

yes, that's where I was also concerned.